### PR TITLE
IsUsedDestination should count any known single-key address

### PR DESCRIPTION
**Pull request by instagibbs**
This plugs the privacy leak detailed at https://github.com/bitcoin/bitcoin/issues/17605, at least for the single-key case.

### CODE REVIEW

**Comment by instagibbs** [comment]
cc @kallewoof since you designed the first iteration

**Comment by achow101** [comment]
Concept ACK

**Comment by kallewoof** [comment]
Concept ACK

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -729,7 +729,22 @@ bool CWallet::IsUsedDestination(const uint256& hash, unsigned int n) const
...
+        assert(srctx->tx->vout.size() >= n);
+        LegacyScriptPubKeyMan* spk_man = GetLegacyScriptPubKeyMan();
+        assert(spk_man != nullptr);
+        for (const auto& keyid : GetAffectedKeys(srctx->tx->vout[n].scriptPubKey, *spk_man)) {
+            WitnessV0KeyHash wpkh_dest(keyid);
+            PKHash pkh_dest(keyid);
+            ScriptHash sh_wpkh_dest(wpkh_dest);
+            if (IsUsedDestination(wpkh_dest) ||
+                    IsUsedDestination(pkh_dest) ||
+                    IsUsedDestination(sh_wpkh_dest)) {
```
Comment (jonatack) [inline src/wallet/wallet.cpp]: Suggestion (if you retouch this or revisit the PR to add tests):
```diff
             if (IsUsedDestination(wpkh_dest) ||
-                    IsUsedDestination(pkh_dest) ||
-                    IsUsedDestination(sh_wpkh_dest)) {
+                IsUsedDestination(pkh_dest) ||
+                IsUsedDestination(sh_wpkh_dest)) {
```

**Review by jonatack — COMMENTED** [review]
Concept ACK on patching this in the interim before descriptor wallets arrive. Edit: Code review, built, ran tests, tested manually.

**Comment by instagibbs** [comment]
Pushed test, updated logic to fix test failure.

Removed a function that resulted in a used footgun.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -719,17 +719,31 @@ void CWallet::SetUsedDestinationState(WalletBatch& batch, const uint256& hash, u
...
 
-bool CWallet::IsUsedDestination(const CTxDestination& dst) const
-{
-    LOCK(cs_wallet);
-    return IsMine(dst) && GetDestData(dst, "used", nullptr);
-}
-
 bool CWallet::IsUsedDestination(const uint256& hash, unsigned int n) const
 {
+    AssertLockHeld(cs_wallet);
```
Comment (promag) [inline src/wallet/wallet.cpp]: nit, also add the EXCLUSIVE_LOCKS_REQUIRED annotation.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -719,17 +719,31 @@ void CWallet::SetUsedDestinationState(WalletBatch& batch, const uint256& hash, u
...
     CTxDestination dst;
     const CWalletTx* srctx = GetWalletTx(hash);
-    return srctx && ExtractDestination(srctx->tx->vout[n].scriptPubKey, dst) && IsUsedDestination(dst);
+    if (srctx) {
+        assert(srctx->tx->vout.size() >= n);
```
Comment (promag) [inline src/wallet/wallet.cpp]: Should be `>`?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -719,17 +719,31 @@ void CWallet::SetUsedDestinationState(WalletBatch& batch, const uint256& hash, u
...
+        LegacyScriptPubKeyMan* spk_man = GetLegacyScriptPubKeyMan();
+        assert(spk_man != nullptr);
+        for (const auto& keyid : GetAffectedKeys(srctx->tx->vout[n].scriptPubKey, *spk_man)) {
+            WitnessV0KeyHash wpkh_dest(keyid);
+            if (IsMine(wpkh_dest) && GetDestData(wpkh_dest, "used", nullptr)) {
```
Comment (promag) [inline src/wallet/wallet.cpp]: Haven't profiled but looks like `GetDestData && IsMine` should be faster?

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: d'oh, I always have to sit there and reason with myself on this :)

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: I just got rid of the `IsMine` check. If a record for `used` exists in the wallet we should trust it.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: done

**Review by kallewoof — APPROVED** [review]
utACK

**Comment by fjahr** [comment]
ACK 9f7eea56f2b0fbc128b83fccab344190b1e6ce54

Reviewed code and tested locally.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: The annotation is supposed to go in the header file.

Comment (instagibbs) [reply src/wallet/wallet.cpp]: is it just a no-op here? I am not a clang wizard.

File: test/functional/wallet_avoidreuse.py [hunk 5]
```diff
@@ -193,7 +198,7 @@ def test_fund_send_fund_send(self):
         '''
         self.log.info("Test fund send fund send")
 
-        fundaddr = self.nodes[1].getnewaddress()
+        fundaddr = self.nodes[1].getnewaddress(label="", address_type="legacy")
```
Comment (achow101) [inline test/functional/wallet_avoidreuse.py]: nit: `label` isn't needed

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: `Attributes must be attached to named declarations` oooohkay fixed

Comment (achow101) [reply src/wallet/wallet.cpp]: From [clang docs](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#frequently-asked-questions):

> Attributes are part of the formal interface of a function, and should always go in the header, where they are visible to anything that includes the header. Attributes in the .cpp file are not visible outside of the immediate translation unit, which leads to false negatives and false positives.

**Review by achow101 — COMMENTED** [review]
ACK 9f7eea56f2b0fbc128b83fccab344190b1e6ce54

File: test/functional/wallet_avoidreuse.py (hunk 5 above)
Comment (kallewoof) [reply test/functional/wallet_avoidreuse.py]: It would be, once address type default is changed, no?

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -2927,7 +2927,7 @@ static UniValue listunspent(const JSONRPCRequest& request)
         CTxDestination address;
         const CScript& scriptPubKey = out.tx->tx->vout[out.i].scriptPubKey;
         bool fValidAddress = ExtractDestination(scriptPubKey, address);
-        bool reused = avoid_reuse && pwallet->IsUsedDestination(address);
+        bool reused = avoid_reuse && pwallet->IsUsedDestination(out.tx->GetHash(), out.i);
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: nit, you could move this line after L2932 - no point in calling `IsUsedDestination` if this address doesn't matter.

**Review by promag — COMMENTED** [review]
Concept ACK.

**Review by meshcollider — COMMENTED** [review]
Code Review ACK 09502452bbbe21bb974f1de8cf53196373921ab9

**Comment by fanquake** [comment]
Being backported in 17792.

**Pull request merged**